controllers/subscription_controller.go

```diff
@@ -97,6 +113,87 @@ func (r *SubscriptionReconciler) Reconcile(ctx context.Context, _ ctrl.Request)
 	return ctrl.Result{}, nil
 }
 
+func getTargetNamespaces(olmPkgRecords []*OlmPkgRecord) []string {
+
+	var namespacesMap = make(map[string]bool)
+	var namespaces []string
+
+	for _, olmPkgRecord := range olmPkgRecords {
+		if !namespacesMap[olmPkgRecord.Namespace] {
+			namespaces = append(namespaces, olmPkgRecord.Namespace)
+		}
+		namespacesMap[olmPkgRecord.Namespace] = true
+	}
+
+	return namespaces
+}
+
+func ensureNamespace(ctx context.Context, cli client.Client, logger logr.Logger, olmPkgRecords []*OlmPkgRecord) error {
+
+	namespacesMap := make(map[string]bool)
+
+	for _, olmPkgRecord := range olmPkgRecords {
+
+		if olmPkgRecord.Namespace == "" || olmPkgRecord.Namespace == OperatorNamespace {
+			logger.Info("skipping namespace creation", "record", olmPkgRecord)
+			continue
+		}
+
+		if !namespacesMap[olmPkgRecord.Namespace] {
+			ns := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: olmPkgRecord.Namespace,
+				},
+			}
+			if err := cli.Create(ctx, ns); err != nil && !errors.IsAlreadyExists(err) {
```

Comment:
any reason not using `CreateOrUpdate`, using `Create` here always makes a call to k8s api while the former go through the cache and let the mutate fn be empty.

Reply:
I am doing get and skip.